Index: libgc-7.1/include/gc_config_macros.h
===================================================================
--- libgc-7.1.orig/include/gc_config_macros.h	2011-03-12 18:35:55.961958955 +0100
+++ libgc-7.1/include/gc_config_macros.h	2011-03-12 18:36:09.169958955 +0100
@@ -102,7 +102,7 @@
 #   define GC_DARWIN_THREADS
 #   define GC_PTHREADS
 # endif
-# if !defined(GC_PTHREADS) && (defined(__FreeBSD__) || defined(__DragonFly__))
+# if !defined(GC_PTHREADS) && (defined(__FreeBSD__) || defined(__DragonFly__) || defined(__FreeBSD_kernel__))
 #   define GC_FREEBSD_THREADS
 #   define GC_PTHREADS
 # endif
@@ -118,6 +118,10 @@
 #   define GC_AIX_THREADS
 #   define GC_PTHREADS
 # endif
+# if defined(__gnu_hurd__)
+#   define GC_GNU_THREADS
+#   define GC_PTHREADS
+# endif
 #endif /* GC_THREADS */
 
 #if defined(GC_THREADS) && !defined(GC_PTHREADS) && !defined(GC_WIN32_THREADS) \
